schedulelambda/src/main/scala/com/gu/notificationschedule/ProcessNotificationScheduleLambda.scala

```diff
@@ -3,39 +3,71 @@ package com.gu.notificationschedule
 }
 
 
-
 class ProcessNotificationScheduleLambda(config: NotificationScheduleConfig, cloudWatch: CloudWatch, notificationSchedulePersistence: NotificationSchedulePersistence) {
-  def this(config: NotificationScheduleConfig, lambdaName: String) = this(config, new CloudWatchImpl(config.stage, lambdaName, AmazonCloudWatchAsyncClientBuilder.defaultClient()), new NotificationSchedulePersistenceImpl(config, AmazonDynamoDBAsyncClientBuilder.defaultClient()) )
+  def this(config: NotificationScheduleConfig, lambdaName: String) = this(config, new CloudWatchImpl(config.stage, lambdaName, AmazonCloudWatchAsyncClientBuilder.defaultClient()), new NotificationSchedulePersistenceImpl(config, AmazonDynamoDBAsyncClientBuilder.defaultClient()))
 
   def this() = this(AppIdentity.whoAmI(defaultAppName = "mobile-notifications-schedule") match {
     case awsIdentity: AwsIdentity => NotificationScheduleConfig(awsIdentity.app, awsIdentity.stage, awsIdentity.stack)
     case _ => throw new IllegalStateException("Not in aws")
   }, "MobileNotificationSchedule")
 
-  val logger = LogManager.getLogger(classOf[ProcessNotificationScheduleLambda])
+  val logger: Logger = LogManager.getLogger(classOf[ProcessNotificationScheduleLambda])
 
   def apply(): Unit = {
-    val notificationsScheduleEntries = notificationSchedulePersistence.query()
+    val timer = cloudWatch.startTimer("lambda")
+    val triedFetch: Try[Seq[NotificationsScheduleEntry]] = Try {
+      notificationSchedulePersistence.query()
+    }
+    val triedAll: Try[Unit] = triedFetch match {
+      case Success(notificationsScheduleEntries) => triggerNotifications(notificationsScheduleEntries)
+      case failure => {
+        cloudWatch.queueMetric("discovery-failed", 1, StandardUnit.Count)
+        failure.map(_ => ())
+      }
+    }
+    triedAll.fold(throwable => timer.fail, unit => timer.succeed)
+    cloudWatch.sendMetricsSoFar()
+  }
+
+  private def triggerNotifications(notificationsScheduleEntries: Seq[NotificationsScheduleEntry]) = {
+    cloudWatch.queueMetric("discovered", notificationsScheduleEntries.size, StandardUnit.Count)
     val nowEpoch = Instant.now().getEpochSecond
-    notificationsScheduleEntries.foreach(notificationSchedulePersistence.write(_, true, nowEpoch))
+    val triedUnits = notificationsScheduleEntries.map(notificationsScheduleEntry => triggerNotification(nowEpoch, notificationsScheduleEntry))
+    val failed = triedUnits.count(_.isFailure)
+    cloudWatch.queueMetric("trigger-failed", failed, StandardUnit.Count)
+    if(failed > 0) {
+      Success(())
+    }
+    else {
+      Failure(new Exception(s"$failed triggers failed out of ${triedUnits.size}"))
+    }
 
-    cloudWatch.sendMetricsSoFar()
+  }
+
+  private def triggerNotification(nowEpoch: Long, notificationsScheduleEntry: NotificationsScheduleEntry) = {
+    Try {
+      logger.warn("TODO: trigger notification {}", notificationsScheduleEntry)
```

Comment:
TODO

Reply:
Yup.
Any placeholder will do, I was trying to avoid doing the whole thing at once (just wanted a lambda wired in first... it's already evolved to include dynamodb, so could do the rest here too).

Reply:
Makes sense then